server/src/main/java/org/opensearch/gateway/PrimaryShardBatchAllocator.java

```diff
@@ -61,50 +62,58 @@ protected FetchResult<TransportNodesListGatewayStartedShards.NodeGatewayStartedS
+            List.of(unassignedShard),
+            Collections.emptyList(),
+            allocation
+        );
+        List<NodeGatewayStartedShard> nodeGatewayStartedShards = adaptToNodeShardStates(unassignedShard, shardsState);
+        return getAllocationDecision(unassignedShard, allocation, nodeGatewayStartedShards, logger);
     }
 
     /**
-     * Build allocation decisions for all the shards present in the batch identified by batchId.
+     * Allocate Batch of unassigned shard  to nodes where valid copies of the shard already exists
      *
-     * @param shards     set of shards given for allocation
-     * @param allocation current allocation of all the shards
-     * @param logger     logger used for logging
-     * @return shard to allocation decision map
+     * @param shardRoutings the shards to allocate
+     * @param allocation    the allocation state container object
      */
-    @Override
-    public HashMap<ShardRouting, AllocateUnassignedDecision> makeAllocationDecision(
-        List<ShardRouting> shards,
-        RoutingAllocation allocation,
-        Logger logger
-    ) {
-        HashMap<ShardRouting, AllocateUnassignedDecision> shardAllocationDecisions = new HashMap<>();
+    public void allocateUnassignedBatch(List<ShardRouting> shardRoutings, RoutingAllocation allocation) {
+        HashMap<ShardId, AllocateUnassignedDecision> ineligibleShardAllocationDecisions = new HashMap<>();
         List<ShardRouting> eligibleShards = new ArrayList<>();
         List<ShardRouting> inEligibleShards = new ArrayList<>();
         // identify ineligible shards
-        for (ShardRouting shard : shards) {
+        for (ShardRouting shard : shardRoutings) {
             AllocateUnassignedDecision decision = getInEligibleShardDecision(shard, allocation);
             if (decision != null) {
+                ineligibleShardAllocationDecisions.put(shard.shardId(), decision);
                 inEligibleShards.add(shard);
-                shardAllocationDecisions.put(shard, decision);
             } else {
                 eligibleShards.add(shard);
             }
         }
-        // Do not call fetchData if there are no eligible shards
-        if (eligibleShards.isEmpty()) {
-            return shardAllocationDecisions;
-        }
+
         // only fetch data for eligible shards
         final FetchResult<NodeGatewayStartedShardsBatch> shardsState = fetchData(eligibleShards, inEligibleShards, allocation);
 
-        // process the received data
-        for (ShardRouting unassignedShard : eligibleShards) {
-            List<NodeGatewayStartedShard> nodeShardStates = adaptToNodeShardStates(unassignedShard, shardsState);
-            // get allocation decision for this shard
-            shardAllocationDecisions.put(unassignedShard, getAllocationDecision(unassignedShard, allocation, nodeShardStates, logger));
+        RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
```

Comment:
please add a note why you need to iterate all the unassigned shards again here

Reply:
Also, any noticable performance regression after the code to iterate all unassigned shards for every batch?

Reply:
Ack. No regression was observed. Measured performance of this fix on a 20 node cluster with 20 K shards and these were the numbers observed:

```
Batch mode disabled:
14: 30 to 14:32 [Red-yellow]
14:32  to 14:41 [yellow - green]
Batch mode enabled:
13:45 to 13:47[ Red-yellow]
13:47 to 13:55 [ yellow-green] 
```

We did slightly better by a minute.